File: solutions/0013/roman_to_integer.c
// Problem 13 - Roman to Integer
// https://leetcode.com/problems/roman-to-integer/

#include <stddef.h>

static inline int get_value(char c) {
  switch (c) {
      case '\0': return 0;
      case 'I': return 1;
      case 'V': return 5;
      case 'X': return 10;
      case 'L': return 50;
      case 'C': return 100;
      case 'D': return 500;
      case 'M': return 1000;
      default: __builtin_unreachable();
  }
}

int romanToInt(char * s){
    
    int num = 0;
    size_t i = 0;
    
    while (s[i])
    {
    	int val  = get_value(s[i]);
    	int next_val = get_value(s[i+1]);
    	
    	if (val < next_val)
    	{
    		num -= val;
    	}
    	else
    	{
    		num += val;
    	}
    	
    	i++;
    }
    
    return num;
}

File: solutions/0013/roman_to_integer.py
# Problem 13 - Roman to Integer
# https://leetcode.com/problems/roman-to-integer/

class Solution:
    def romanToInt(self, s: str) -> int:
        roman_num = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        prev_val = 0
        out = 0
        
        # Iterate over the reversed string
        for char in s[::-1]:
            val = roman_num[char]
            if val < prev_val:
                out -= val
            else:
                out += val
            prev_val = val
        
        return out